<launch>

	<arg name="stdout" default="log" />
    <arg name="marker_size" value="0.2"/>
    <!--<arg name="delta_x_marker_circle" value="0.25"/>-->
    <!--<arg name="delta_y_marker_circle" value="0.075"/>-->
    <!--<arg name="delta_x_centers" value="0.45"/>-->
    <!--<arg name="delta_y_centers" value="0.4"/>-->
    <arg name="delta_x_marker_circle" value="0.30"/>
    <arg name="delta_y_marker_circle" value="0.15"/>
    <arg name="delta_x_centers" value="0.5"/>
    <arg name="delta_y_centers" value="0.4"/>

    <node pkg="tf" type="static_transform_publisher" name="camera_ros_tf" args="0 0 0 -1.57079632679 0 -1.57079632679 rotated_stereo_camera stereo_camera 10"/>

    <node pkg="velo2cam_calibration" type="mono_qr_pattern" name="mono_qr_pattern" output="$(arg stdout)">
		<param name="image_topic" value="/stereo_camera/left/image_rect_color" />
		<param name="cinfo_topic" value="/stereo_camera/left/camera_info" />
        <param name="marker_size" value="$(arg marker_size)"/>
        <param name="delta_x_marker_circle" value="$(arg delta_x_marker_circle)"/>
        <param name="delta_y_marker_circle" value="$(arg delta_y_marker_circle)"/>
        <param name="delta_x_centers" value="$(arg delta_x_centers)"/>
        <param name="delta_y_centers" value="$(arg delta_y_centers)"/>
	</node>

    <!-- PCL PREPROCESSING -->
    <!-- PCL nodelet manager -->
    <node pkg="nodelet" type="nodelet" name="pcl_manager" args="manager" output="$(arg stdout)"/>

    <node pkg="nodelet" type="nodelet" name="pass_through_z_velo" args="load pcl/PassThrough pcl_manager" output="$(arg stdout)">
        <remap from="~input" to="/velodyne_points" />
        <remap from="~output" to="/velodyne/z_filtered" />
        <rosparam>
            filter_field_name: y
            filter_limit_min: -100
            filter_limit_max: 100
            filter_limit_negative: False
            max_queue_size: 1
            keep_organized: False
        </rosparam>
    </node>

    <node pkg="nodelet" type="nodelet" name="pass_through_x_velo" args="load pcl/PassThrough pcl_manager" output="$(arg stdout)">
        <remap from="~input" to="/velodyne/z_filtered" />
        <remap from="~output" to="/velodyne/xy_filtered" />
        <rosparam>
            filter_field_name: x
            filter_limit_min: 0
            filter_limit_max: 100
            filter_limit_negative: False
            max_queue_size: 1
            keep_organized: False
        </rosparam>
    </node>
    <node pkg="velo2cam_calibration" type="laser_pattern" name="laser_pattern" output="$(arg stdout)">
        <remap from="~cloud1" to="/velodyne/xy_filtered"/>
        <rosparam>
            x: 0
            y: 0
            z: 1
            angle_threshold: 0.55
            circle_radius: 0.12
            passthrough_radius_min: 1.0
            passthrough_radius_max: 6.0
            centroid_distance_min: 0.15
            centroid_distance_max: 0.8
            rings_count: 64
        </rosparam>
    </node>

	<node pkg="velo2cam_calibration" type="velo2cam_calibration" name="velo2cam_calibration" output="screen">
		<remap from="~cloud1" to="/laser_pattern/centers_cloud"/>
		<remap from="~cloud2" to="/mono_qr_pattern/centers_cloud"/>
		<remap from="~cloud3" to="/velodyne_points"/>
		<param name="radius" value="0.12"/>
		<param name="centers_distance" value="0.3"/>
        <param name="delta_x_centers" value="$(arg delta_x_centers)"/>
        <param name="delta_y_centers" value="$(arg delta_y_centers)"/>
        <param name="is_sensor1_cam" value="false"/>
        <param name="is_sensor2_cam" value="true"/>
	</node>
</launch>
